src/ext/java/org/opentripplanner/ext/fares/impl/DefaultFareService.java

```diff
@@ -102,21 +102,81 @@ public ItineraryFares calculateFares(Itinerary itinerary) {
     if (fareLegs.isEmpty()) {
       return null;
     }
+    var fareLegsByFeed = fareLegs
+      .stream()
+      .collect(Collectors.groupingBy(leg -> leg.getAgency().getId().getFeedId()));
+    var fareRulesByTypeAndFeed = fareRulesPerType
+      .entrySet()
+      .stream()
+      .collect(
+        Collectors.toMap(
+          Map.Entry::getKey,
+          rules ->
+            rules
+              .getValue()
+              .stream()
+              .collect(Collectors.groupingBy(rule -> rule.getFareAttribute().getId().getFeedId()))
+        )
+      );
 
     ItineraryFares fare = ItineraryFares.empty();
     boolean hasFare = false;
-    for (Map.Entry<FareType, Collection<FareRuleSet>> kv : fareRulesPerType.entrySet()) {
-      FareType fareType = kv.getKey();
-      Collection<FareRuleSet> fareRules = kv.getValue();
-      // Get the currency from the first fareAttribute, assuming that all tickets use the same currency.
-      if (fareRules.size() > 0) {
-        Currency currency = Currency.getInstance(
-          fareRules.iterator().next().getFareAttribute().getCurrencyType()
+    for (FareType fareType : fareRulesPerType.keySet()) {
+      List<FareComponent> components = new ArrayList<>();
+      List<Money> fares = new ArrayList<>();
+      ItineraryFares currentFare = ItineraryFares.empty();
+      boolean legWithoutRulesFound = false;
+      for (String feedId : fareLegsByFeed.keySet()) {
+        var fareRules = fareRulesByTypeAndFeed.get(fareType).get(feedId);
+
+        // Get the currency from the first fareAttribute, assuming that all tickets use the same currency.
+        if (fareRules != null && fareRules.size() > 0) {
+          Currency currency = Currency.getInstance(
+            fareRules.iterator().next().getFareAttribute().getCurrencyType()
+          );
+          hasFare =
+            populateFare(currentFare, currency, fareType, fareLegsByFeed.get(feedId), fareRules) ||
+            hasFare; // Other feeds might still have fare for some legs
+
+          components.addAll(currentFare.getComponents(fareType));
+          fare.addFare(fareType, currentFare.getFare(fareType));
+          fares.add(currentFare.getFare(fareType));
+
+          // If all the legs are from one feed, consider itinerary products
+          if (fareLegs.equals(fareLegsByFeed.get(feedId))) fare.addItineraryProducts(
+            currentFare.getItineraryProducts()
+          );
+        } else {
+          legWithoutRulesFound = true;
+        }
+      }
+
+      fare.addFareComponent(fareType, components);
+
+      // No fares will be discovered after this point
+      if (!hasFare) {
+        return null;
+      }
+
+      // Accumulate the final price of the fare or indicate that no final fare could be found
+      if (legWithoutRulesFound) {
+        fare.addFare(
+          fareType,
+          Money.ofFractionalAmount(fares.get(0).currency(), Float.POSITIVE_INFINITY)
```

Comment:
This is quite dubious. You're adding an infinite fare when you have a leg without rules?

Reply:
Yes, it looks a bit funny but that seems to be how the fare is set for similar single feed cases in the performSearch method. I could of course change that to  directly give the -0.01 that seems to be the current way to indicate that some legs do not have fares. Or do you maybe have some other ideas on how this should be handled?

Reply:
Sure, but we are not in the `performSearch` method. The positive infinity is directly returned to the client.

If we want -1 cent then we need to return it here.

Also, we need lots and lots of tests that document what exactly the expected return values are since it's very confusing.

Reply:
Ok, I'll change that infinity to  the -0.01.  I added two test cases for the new multifeed logic. Is there some pre-existing functionality that I should add tests for in this PR?

Reply:
There _is_ lots of functionality that we should test but knowledge about what the expectations are has been lost over time. The best we can do is to write tests when we add new features or find bugs.